website/blog/2022-08-17-managing-surrogate-keys-in-dbt.md

```diff
@@ -0,0 +1,228 @@
+
+```sql
+# in models/reports/daily_user_orders.sql
+with 
+
+orders as (
+	select * from {{ ref('fct_orders') }} 
+),
+
+agg as (
+
+	select 
+		date_trunc(day, order_date) as report_date 	
+		user_id, 
+		count(*) as total_orders
+  from orders
+  group by 1,2
+
+),
+
+final as (
+	
+	select
+		{{ dbt_utils.surrogate_key([
+				'report_date', 
+				'user_id'
+			])
+		}} as unique_key, 
+		*
+	from agg
+
+)
+
+select * from final
+```
+
+Using hashed keys makes our transformations idempotent — every dbt run results in the same exact outputs every time. I can safely delete all my non-source objects in my warehouse, execute a dbt run and be right back where I started (though I wouldn’t necessarily recommend this 😅).  
+
+The analytical warehouses we use now no longer have the same constraints that traditional warehouses had  — joins on strings aren’t notably less performant than those on integers, and storing slightly larger values in the surrogate key column is peanuts given the relative cost of storage on these platforms. This strategy also removes the need for tight coupling of transformations to propagate the surrogate key values across our project — anywhere the inputs for the surrogate keys are present, the hashing function produces the same keys, so we can take advantage of parallel processing in our warehouse and avoid the bottlenecks we had before. 
+
+### Caveats
+
+This strategy is not without its caveats either!
+
+- **Collisions -** Although it is exceedingly rare, it is possible for two different sets of inputs to produce the same outputs, causing erroneous duplicate records in your dataset. Using an MD5 hash (the default for the `dbt_utils.surrogate_key` macro), you have a 50% of a collision when you get up to 2^64 records (1.84 x 10E19 aka a whole lot of data). While very unlikely, it’s certainly something to consider for truly massive datasets.
+- **Performance -** As mentioned above, hashed keys result in long string-type values. On truly massive datasets, this can cause some performance issues. Unlike MIISKs, string values can’t be easily partitioned to improve query performance.
```

Comment:
This section is somewhat ominous haha. Do you have recommendations for huge datasets using hashed surrogate keys? Ex. Using incremental/table materializations.

Reply:
i've sometimes recommended using different datatypes as in the next bullet point! I think i can add a sentence here to make that clearer.

Reply:
the collisions piece i think is mostly unavoidable unfortunately 😬

Reply:
Oh sorry should have been more specific! The section on Performance seems a little ominous to me!!

Reply:
I think you could call out using other hashing algorithms (SHA256 or something) in the Collisions section, but again, an md5 collision is so very unlikely so it could be overkill

Reply:
I rearranged them and softened the language in performance a bit! let me know what you think!

Reply:
Looking great to me :)